### PR TITLE
In problem manual grader

### PR DESCRIPTION
**Pull request by drgrice1**
This adds a new manual problem grader that is available for all problems and works in gateway quizzes as well.  To accommodate grading, a student nav is added to the top of the problem and gateway quiz pages.

There is code clean up of the GatewayQuiz.pm, ProblemGrader.pm, and ProblemSet.pm files in the first commit.  Mostly fixing indentation and removing white space at line ends.  All the essential work is in the remaining commits.

I am open to suggestions for improvement, but this is generally working well, and is a nice feature to have.

### CODE REVIEW

**Comment by Alex-Jordan** [comment]
Glenn, this is great. I'm eager to apply this to our server.

I tested as follows. (TLDR: it worked as intended)

* Switched to this branch and restarted apache.
* Created a testing course and made a problem set with two problems. One "normal", one with an essay answer.
* Created a student and assigned the set to them. Student submitted an incorrect answer to the normal exercise and an answer to the essay exercise.
* Signed in as instructor. Clicked my way to acting as the student. Visited the problems. Checkboxed "ProblemGrader" and clicked Check Answers.
* Assigned a score for one problem and left a comment. Clicked Next Problem and did the same. Scores are reflected in Student Progress.
* Signed in as student, and I see the score and comments when visiting each problem.
* Copied the problem set and made it a GW quiz. Student took the quiz.
* As instructor, acting as student, visited "testQuiz (test 1)". Checkboxed "Show problem graders" and clicked Check Test. Gave feedback, changed scores, and saved. Grades reflected in Student Progess. Student can see updated scores and feedback.


I'd like to try to proofread the code edits in the 15 affected files (and 2 files are new). Without being familiar with what was needed for this, it's hard to know if I should test other things for unintended consequences. But if you have these commits on your server, that is at least a good field test.

**Comment by Alex-Jordan** [comment]
There are to math4. Our server has a custom theme copied from math4 and modified. I will need to update the theme files to reflect this. 

My point is that some educating will be needed for anyone with a house theme who upgrades to 2.16. We should make a note in the release notes, and also be ready for reports of something "not working" to appear in the forum.

**Review by Alex-Jordan — APPROVED** [review]
It worked, and my scan of file diffs didn't raise concerns. We should do something in the release notes to make it clear to those with a custom theme that they need to update their theme files as math4 was updated.

Of course, I did not really scan the large diffs closely. Some of the files have a lot of white space edits to make them conform to standards. It just makes it hard to focus on the meaningful edits with the feature here. I wonder if we could take a deep breath at some point and tackle code formatting all at once so it won't get in the way moving forward. 

Maybe someone else should test this too. But I'm also comfortable just merging it if time passes and no one gets around to testing.

**Comment by drgrice1** [comment]
@Alex-Jordan:  Thanks for your review.

It helps to use the "Hide whitespace changes" setting, although not that much.  Also all of those changes are in the first commit, and that is all that is in that commit.  So if you do "git diff" from that commit in this pull request to the last, you will only see the relevant changes.  You can ignore all changes to lib/WeBWorK/ContentGenerator/Instructor/ProblemGrader.pm.  There are no relevant changes to that file, just clean up.  Perhaps I should have not done the clean up, but I was having a hard time dealing with the mess and understanding the code to make the needed changes for this pull request.  The GatewayQuiz.pm file in particular needs this clean up.  It is one of the worst in the entire WeBWorK code base for readability.

As to the math4 changes, note that the changes here are not the only changes to the theme since WeBWorK 2.15.  I think a comment should be added to the upgrade documentation for every release to check custom theme compatibility with potential changes to the provided theme and the system.  I use a custom theme, and I had to make changes to that theme when upgrading from WeBWorK 2.14 to 2.15 also.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 1]
```diff
@@ -65,135 +66,133 @@ sub templateName {
...
+		($authz->hasPermissions($User->user_id, "view_hidden_work") || $canShowScores));
+}
+
+sub can_showProblemGrader {
+	my ($self, $User, $PermissionLevel, $EffectiveUser, $Set, $Problem) = @_;
+	my $authz = $self->r->authz;
+
+	return ($authz->hasPermissions($User->user_id, "access_instructor_tools") &&
+		$authz->hasPermissions($User->user_id, "score_sets"));
 }
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:123]: Added method for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 2]
```diff
@@ -412,12 +406,27 @@ sub handle_input_colors {
...
+
+	if ($userPastAnswerID) {
+		my $userPastAnswer = $db->getPastAnswer($userPastAnswerID);
+		return $userPastAnswer->comment_string;
+	}
+
+	return "";
 }
 
 ################################################################################
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:432]: Added method for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 3]
```diff
@@ -975,80 +928,78 @@ sub pre_header_initialize {
...
-	     useWirisEditor     => $User->useWirisEditor ne '' ? $User->useWirisEditor : $ce->{pg}->{options}->{useWirisEditor},
-	     useMathQuill       => $User->useMathQuill ne '' ? $User->useMathQuill : $ce->{pg}->{options}->{useMathQuill},
-	     );
+	my %want = (
+		showOldAnswers     => $User->showOldAnswers ne '' ?
+			$User->showOldAnswers : $ce->{pg}->{options}->{showOldAnswers},
+		# showProblemGrader implies showCorrectAnswers.  This is a convenience for grading.
+		showCorrectAnswers => ($r->param("showCorrectAnswers") || $r->param('showProblemGrader') ||
+			$ce->{pg}->{options}->{showCorrectAnswers}) && ($submitAnswers || $checkAnswers),
+		showProblemGrader  => $r->param('showProblemGrader') || 0,
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:943]: Changed/added in this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 4]
```diff
@@ -975,80 +928,78 @@ sub pre_header_initialize {
...
-	     useMathView              => $self->can_useMathView(@args),
-	     useWirisEditor           => $self->can_useWirisEditor(@args),
-	     useMathQuill           => $self->can_useMathQuill(@args)
-	     );
+	my @args = ($User, $PermissionLevel, $EffectiveUser, $set, $Problem, $tmplSet);
+	my $sAns = $submitAnswers ? 1 : 0;
+	my %can = (
+		showOldAnswers        => $self->can_showOldAnswers(@args),
+		showCorrectAnswers    => $self->can_showCorrectAnswers(@args, $sAns),
+		showProblemGrader     => $self->can_showProblemGrader(@args),
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:976]: Added in this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 5]
```diff
@@ -1183,7 +1129,7 @@ sub pre_header_initialize {
 
 sub head {
 	my ($self) = @_;
-	return if !defined($self->{ra_pg_results});
+	return "" if !defined($self->{ra_pg_results});
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:1132]: Minor mistake of mine from a previous pull request, fixed here.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 6]
```diff
@@ -1194,31 +1140,119 @@ sub head {
...
+					data_toggle => "tooltip", data_placement => "top",
+					title => "$nextTest->{displayName} (test $nextTest->{setVersion})",
+					class => "nav_button student-nav-button"
+				}, $r->maketext("Next Test"))
+			: CGI::span({ class => "gray_button" }, $r->maketext("Next Test")),
+			CGI::end_div()
+		);
+	}
+
+	return "";
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:1255]: This is a big part of this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 7]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
-								  $pg->{flags}->{showPartialCorrectAnswers} && $canShowProblemScores,
-								  $canShowProblemScores, 1);
-					
-				} elsif ( $will{checkAnswers} ) {
+					$resultsTable =
+					$self->attemptResults($pg, 1, $will{showCorrectAnswers},
+						$pg->{flags}->{showPartialCorrectAnswers} && $canShowProblemScores,
+						$canShowProblemScores, 1);
+
+				} elsif ($will{checkAnswers} || $will{showProblemGrader}) {
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2100]: Change for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 8]
```diff
@@ -42,10 +41,12 @@ use WeBWorK::Debug;
 use WeBWorK::ContentGenerator::Instructor qw(assignSetVersionToUser);
 use WeBWorK::Authen::LTIAdvanced::SubmitGrade;
 use WeBWorK::Utils::AttemptsTable;
+use WeBWorK::ContentGenerator::Instructor::SingleProblemGrader;
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:44]: Change in code for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 9]
```diff
@@ -975,80 +928,78 @@ sub pre_header_initialize {
...
-	     recordAnswers      => 0,
-	     checkAnswers       => 0,
-	     useMathView        => 0,
-	     useWirisEditor     => 0,
-	     useMathQuill     => 0,
-	     );
+	my %must = (
+		showOldAnswers     => 0,
+		showCorrectAnswers => 0,
+		showProblemGrader  => 0,
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:960]: Added for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 10]
```diff
@@ -1310,29 +1340,28 @@ sub body {
...
 
-			$message .= "$errorNum. " if ( @pg_errors > 1 );
+			$message .= "$errorNum. " if (@pg_errors > 1);
 			$message .= $_->{message} . CGI::br() . "\n";
 
-			$context .= CGI::p((@pg_errors > 1? "$errorNum.": '') . 
-					   $_->{context} ) . "\n\n" . 
-					   CGI::hr() . "\n\n";
+			$context .= CGI::p((@pg_errors > 1? "$errorNum.": '') .
+				$_->{context}) . "\n\n" . CGI::div({ class => 'gwDivider' }, "") . "\n\n";
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:1355]: Minor change here.  hr's should not be used in this way.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 11]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
+				print CGI::h2($r->maketext("Problem [_1].",$problemNumber)), $recordMessage;
+
+				my $instructor_comment = $self->get_instructor_comment($problems[$probOrder[$i]]);
+				if ($instructor_comment) {
+					print CGI::start_div({ id => "answerComment", class => "answerComments" });
+					print CGI::b($r->maketext("Instructor Comment:")),  CGI::br();
+					print CGI::escapeHTML($instructor_comment);
+					print CGI::end_div();
+				}
+
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2126]: Change for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 12]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
 
-				print $resultsTable if $resultsTable; 
+				print $resultsTable if $resultsTable;
+
+				# Initialize the problem graders for the problem.
+				if ($self->{will}{showProblemGrader}) {
+					my $problem_grader = new WeBWorK::ContentGenerator::Instructor::SingleProblemGrader(
+						$self->r, $pg, $problems[$probOrder[$i]]);
+					$problem_grader->insertGrader;
+				}
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2139]: Change for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 13]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
-				# finally, store the problem status for 
+				# finally, store the problem status for
 				#    continued attempts recording
 				my $pNum = $probOrder[$i] + 1;
-				print CGI::hidden({-name=>"probstatus$pNum",
-						   -value=>$probStatus[$probOrder[$i]]});
+				print CGI::hidden({-name=>"probstatus$pNum", -value=>$probStatus[$probOrder[$i]]});
 
-				print "\n", CGI::hr(), "\n";
+				print "\n", CGI::div({ class => 'gwDivider' }, ""), "\n";
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2146]: Again with the hr abuse.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 14]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
-# 	    print CGI::hidden({-name=>$probid, -value=>$probval}), "\n";
+				print CGI::hidden({-name=>"probstatus$pNum", -value=>$probStatus[$probOrder[$i]]});
 			}
 		}
 
 		$self->handle_input_colors;
 
 		print CGI::div($jumpLinks, "\n");
-		print "\n",CGI::hr(), "\n";
+		print "\n", CGI::div({ class => 'gwDivider' }, ""), "\n";
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2171]: hr abuse again.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 15]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
 
 		if ($can{showCorrectAnswers}) {
-			print CGI::checkbox(-name   =>"showCorrectAnswers",
-#				-checked => $will{showCorrectAnswers},
-					    -checked=>$want{showCorrectAnswers},
-					    -label  =>$r->maketext("Show correct answers"),
-					    );
+			print CGI::checkbox(
+				-name    => "showCorrectAnswers",
+				-checked => $want{showCorrectAnswers} || $want{showProblemGrader},
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2176]: Change for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 16]
```diff
@@ -1908,376 +1885,370 @@ sub body {
...
+				-checked => $will{showSolutions},
+				-label   => $r->maketext("Show Solutions"),
+			);
+		}
+		if ($can{showProblemGrader}) {
+			print CGI::checkbox(
+				-name    => "showProblemGrader",
+				-checked => $want{showProblemGrader},
+				-label   => $r->maketext("Show problem graders"),
+			);
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2192]: Change for this pull request.

File: lib/WeBWorK/ContentGenerator/GatewayQuiz.pm [hunk 17]
```diff
@@ -2384,46 +2352,63 @@ sub output_JS{
...
-	print CGI::start_script({type=>"text/javascript", src=>"$site_url/js/vendor/other/knowl.js"}),CGI::end_script();
+
+	print CGI::start_script({type=>"text/javascript",
+			src=>"$site_url/js/vendor/other/knowl.js"}),CGI::end_script();
+
+	# This is for the problem grader
+	if ($self->{will}{showProblemGrader}) {
+		print CGI::start_script({type=>"text/javascript", src=>"$site_url/js/apps/ProblemGrader/problemgrader.js"}),
+			CGI::end_script();
+	}
```
Comment (drgrice1) [inline lib/WeBWorK/ContentGenerator/GatewayQuiz.pm:2401]: Added for this pull request.

**Review by drgrice1 — COMMENTED** [review]
I have marked all of the changes of consequence to the file GatewayQuiz.pm to facilitate review of this pull request.
You can ignore all changes to lib/WeBWorK/ContentGenerator/Instructor/ProblemGrader.pm and lib/WeBWorK/ContentGenerator/ProblemSet.pm b/lib/WeBWorK/ContentGenerator/ProblemSet.pm.  All of those changes are code clean up (that I probably shouldn't have done here).

**Comment by Alex-Jordan** [comment]
I've merged this into my production branch. I normally wouldn't be so cavalier, but being able to easily review and adjust quiz scores (where only one attempt is typically allowed, and students cannot self-correct typos and bad syntax) is worth it. So we are adding to the testing of this.

**Comment by Alex-Jordan** [comment]
> I am open to suggestions for improvement, but this is generally working well, and is a nice feature to have.

Could the grader be tied in to the results table when the results table is present? In the extreme, I visited one instructor's quiz that shows multiple problems, and each problem had around 8 parts, with answer blanks arranged in an ordered list (so, also laid out vertically). So there is a lot of vertical scrolling between the answer blanks (to be sure of what the student typed), the results table, and the grader. And manually discerning in the grader which answer is "Answer 5" and so on.

I imagine merging the results table and grader is a tall order, and if so, please just consider this a random idea tossed out.

**Comment by drgrice1** [comment]
I am sure that it would be possible to move the grader into the results table.  I am not sure about that though.  The results table already gets rather wide in many cases.  Adding even more into that could be a mess.

**Review by glarose — APPROVED** [review]

**Comment by drgrice1** [comment]
Thanks @glarose.  Now that I think about it, I probably should have requested your review.

**Comment by somiaj** [comment]
Hello,

I have switched to using this drop down menu, but I have 80 students in a single class, so the drop down menu extends too far down the page. Adding a max-height and overflow-y to the menu makes it so a scroll bar appears as opposed to showing the full menu. Adding the following css makes the menu far easier to use in large classes. Unsure if you want to add this to the dropdown-menu class (might affect other dropdown menus which could be useful) or create a new class for just this menu.

```css
.dropdown-student-list {
    max-height: 200px;
    overflow-y: auto;
}
```

**Comment by drgrice1** [comment]
I actually prefer that the whole menu be shown.  You can scroll down to see the menu without that css style.  Selecting a student reloads the page, so you don't need to scroll back up.  Note that at most 100 students will be shown in any case (at most 50 before and 50 after the currently selected student).

You can add that style to your math4-overrides.css file to get it for your server.  I recommend the following to avoid the possibility of interfering with other bootstrap dropdowns:
```css
.problem-nav .user-nav .dropdown-menu {
	max-height: 200px;
	overflow-y: auto;
}
```

I am not sure where you got the css class ```dropdown-student-list```.  That class isn't used, and doesn't work.

If the general consensus is that this approach is better, then we can put this into the math4.css file (and I can override it back with my math4-overrides.css file).  I would suggest more than 200px though.  Something like 500px looks good to me.

**Comment by somiaj** [comment]
I just made `dropdown-student-list` up, as a a suggestion for a new class to only affect that single menu, vs modify the .dropdown-menu class. I did fail to notice that it was wrapped in the .user-nav class div, though I'm not seeing the .problem-nav class wrapping the menu for the gateway quiz, this seems to be homework specific.

Anyways, I can just make the changes locally, I was just thinking others might like it (in my case, the menu extended a ways below the bottom of the problem, and I would rather just scroll the menu than the whole page). As for 200px vs 500px, it is a matter of taste. I ended up using, which works for both homework and gateway quizzes.

```css
.user-nav .student-nav-selector .dropdown-menu {
        max-height: 200px;
        overflow-y: auto;
}
```

**Comment by drgrice1** [comment]
Yeah, I forgot to check gateway quizzes.  They won't have the problem-nav class.

I suggest using vh for height instead of pixels.  Here is the best I have found:
```css
.user-nav .student-nav-selector .dropdown-menu {
	max-height: 60vh;
	overflow-y: auto;
}
```

**Comment by drgrice1** [comment]
I don't like the way that the scrollbar messes with the appearance of the dropdown though.  It kills the rounded corners on the right.

**Comment by drgrice1** [comment]
I added my version of the css to this pull request.

**Comment by somiaj** [comment]
> 
> Note that at most 100 students will be shown in any case (at most 50 before and 50 after the currently selected student).
> 

Any reason for this restriction? I just ran across an instance I was at a student near the top of my roster, so I had to click some student near the bottom, to get a new page with a menu that extend to a student near the end I was trying to access (note this is way better than the old method, more just curious why this restriction is included).

**Comment by drgrice1** [comment]
The reason is to prevent a massive amount of data from being processed and sent and speed page loading.  I tested this with a course that had more than 5000 students, and the amount of time that it took for the page to load if all students were sent at once was to much.  That would happen on each page change or student change.  Every single time you load the page.  I went with a simple approach of at most 50 before and 50 after.

Note that the classlist editor has limits also.  Although, it just shows 0 students if the limits are exceeded.

In addition, it was to prevent the dropdown from getting too long.  I guess with the css style limiting the height of the dropdown to 60% of the viewport height, we could increase the limits a bit say to perhaps 100 before and after.  So at most 200 in the list.  The classlist editor's limit is 200 also.

**Comment by somiaj** [comment]
> 
> The reason is to prevent a massive amount of data from being processed and sent and speed page loading.

Thanks for the explanation. Would it be possible to have a more consistent menu length in larger classes, doing 100 above and below, could result in a list being from 101 to 200 students depending on which student currently selected, or would it add to much complexity/computation power to adjust the up/down to always give the same number of returns in classes > 200 with your new update?

**Comment by drgrice1** [comment]
The last commit changed that... Yeah.

**Comment by Alex-Jordan** [comment]
I last tested this on Jan 27. Since then there have been many commits and some force pushes. It still passes my simple testing. 

Gavin approved this at one point. I think this is a high-value new feature, especially in pandemic times with more online quizzes, more need to manually review student answers, and more essay question answers in place of turning something in on paper in person. It's also a big diff. Could one more person review this and then let's get it merged before Glenn tweaks it more :)

**Comment by drgrice1** [comment]
The only changes since January have been cosmetic and merge fixing.

**Review by drdrew42 — APPROVED** [review]
Works as advertised -- being able to navigate between students is a phenomenal addition. Excellent work.

**Comment by Alex-Jordan** [comment]
This is in production with Glenn's server, my server, and at least one more, with no known issues. And with 2 approvals, I'm calling it and merging. It's big enough that it testing should continue on other things with this already present.

**Pull request merged**